SpatialGDK/Source/SpatialGDK/Private/Interop/Connection/SpatialEventTracer.cpp

```diff
@@ -0,0 +1,288 @@
+		{
+			MaxFileSize = Settings->MaxEventTracingFileSizeBytes;
+			Enable(WorkerId);
+		}
+	}
+}
+
+SpatialEventTracer::~SpatialEventTracer()
+{
+	if (IsEnabled())
+	{
+		Disable();
+	}
+}
+
+TOptional<Trace_SpanId> SpatialEventTracer::TraceEvent(const FEventMessage& EventMessage, const UStruct* Struct,
+													   const TArray<worker::c::Trace_SpanId>& Causes)
+{
+	if (!IsEnabled())
+	{
+		return {};
+	}
+
+	Trace_SpanId CurrentSpanId;
+	if (Causes.Num() > 0)
+	{
+		CurrentSpanId = Trace_EventTracer_AddSpan(EventTracer, Causes.GetData(), Causes.Num());
+	}
+	else
+	{
+		CurrentSpanId = Trace_EventTracer_AddSpan(EventTracer, nullptr, 0);
+	}
+
+	Trace_Event TraceEvent{ CurrentSpanId, /* unix_timestamp_millis: ignored */ 0, EventMessage.GetMessage(), EventMessage.GetType(),
+							nullptr };
+	if (!Trace_EventTracer_ShouldSampleEvent(EventTracer, &TraceEvent))
+	{
+		return {};
+	}
+
+	Trace_EventData* EventData = Trace_EventData_Create();
+
+	auto AddTraceEventStringField = [EventData](const FString& KeyString, const FString& ValueString) {
+		auto KeySrc = StringCast<ANSICHAR>(*KeyString);
+		const ANSICHAR* Key = KeySrc.Get();
+		auto ValueSrc = StringCast<ANSICHAR>(*ValueString);
+		const ANSICHAR* Value = ValueSrc.Get();
+		Trace_EventData_AddStringFields(EventData, 1, &Key, &Value);
+	};
+
+#if ENGINE_MINOR_VERSION >= 25
+	using UnrealProperty = FProperty;
+#else
+	using UnrealProperty = UProperty;
+#endif
+
+	for (TFieldIterator<UnrealProperty> It(Struct); It; ++It)
+	{
+		UnrealProperty* Property = *It;
+
+		FString VariableName = Property->GetName();
+		const void* Value = Property->ContainerPtrToValuePtr<uint8>(&EventMessage);
+
+		check(Property->ArrayDim == 1); // Arrays not handled yet
+
+#if ENGINE_MINOR_VERSION >= 25
+		if (FStrProperty* StringProperty = CastField<FStrProperty>(Property))
+#else
+		if (UStrProperty* StringProperty = Cast<UStrProperty>(Property))
+#endif
+		{
+			AddTraceEventStringField(VariableName, StringProperty->GetPropertyValue(Value));
+		}
+#if ENGINE_MINOR_VERSION >= 25
+		else if (FObjectProperty* ObjectProperty = CastField<FObjectProperty>(Property))
+#else
+		else if (UObjectProperty* ObjectProperty = Cast<UObjectProperty>(Property))
+#endif
+		{
+			UObject* Object = ObjectProperty->GetPropertyValue(Value);
+			if (Object)
+			{
+				AddTraceEventStringField(VariableName, Object->GetName());
+				if (AActor* Actor = Cast<AActor>(Object))
+				{
+					FString KeyString = VariableName + TEXT("Position");
+					FString ValueString = Actor->GetTransform().GetTranslation().ToString();
+					AddTraceEventStringField(KeyString, ValueString);
+				}
+			}
+			else
+			{
+				AddTraceEventStringField(VariableName, "Null");
+			}
+		}
+		else // Default
+		{
+			FString StringValue;
+			Property->ExportTextItem(StringValue, Value, NULL, NULL, PPF_None);
+			AddTraceEventStringField(VariableName, StringValue);
+		}
+
+		// else if (UEnumProperty* EnumProperty = Cast<UEnumProperty>(Property))
```

Comment:
Can probably remove my comments here, the default fallback handles most types well